src/BotwinExtensions.cs

```diff
@@ -35,23 +37,58 @@ public static IApplicationBuilder UseBotwin(this IApplicationBuilder builder, Bo
             ApplyGlobalAfterHook(builder, options);
 
             var routeBuilder = new RouteBuilder(builder);
+            var systemRoutes = new List<(string verb, string route)>();
 
             //Create a "startup scope" to resolve modules from
             using (var scope = builder.ApplicationServices.CreateScope())
             {
+                var modules = scope.ServiceProvider.GetServices<BotwinModule>();
+
                 //Get all instances of BotwinModule to fetch and register declared routes
-                foreach (var module in scope.ServiceProvider.GetServices<BotwinModule>())
+                foreach (var module in modules)
                 {
                     var moduleType = module.GetType();
 
                     foreach (var route in module.Routes.Keys)
                     {
                         routeBuilder.MapVerb(route.verb, route.path, CreateRouteHandler(route, moduleType));
+
+                        var strippedPath = route.path.EndsWith("/") ? route.path.Substring(0, route.path.Length - 1) : route.path;
+                        systemRoutes.Add((route.verb, "/" + strippedPath));
                     }
                 }
+
+                builder.UseRouter(routeBuilder.Build());
+
+                return builder.Use((ctx, next) => GetMethodNotAllowedHandler(ctx, next, systemRoutes));
             }
+        }
+
+        /// <summary>
+        /// This method is only called if it's a valid 404 or 405
+        /// </summary>
+        /// <param name="context"></param>
+        /// <param name="next"></param>
+        /// <param name="systemRoutes"></param>
+        /// <returns></returns>
+        private static async Task GetMethodNotAllowedHandler(HttpContext context, Func<Task> next, IEnumerable<(string verb, string route)> systemRoutes)
+        {
+            //Call the final pipeline which gets ASP.Net Core status code, usually a 404 in this case
+            await next();
```

Comment:
Why are you continuing the pipeline here at the beginning? Shouldn’t this handler already be kind of the final one in the pipeline? Do you need to know that it produced a 404? Wouldn’t it be safe to assume that if this handler runs after the routing middleware, then a route was not found? (again: I have no actual idea about the routing middleware)

Reply:
Oddly before that is called the response status code is 200, if you call it next it's then 404 from the router. Not sure how because as you say I'd expect it to be already at the end of the pipeline

Reply:
I don’t think you actually need to check the status code here though. This should not run if the router succeeded in the pipeline step before.

Reply:
Interestingly if I remove the `next()` call it will always be 200 on a route not defined in Botwin. If I take the GetMethodNotAllowedHandler out completely I get a 404 so I assume there must be one last pipeline called somewhere